collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+The core issues lies in the fact that NIR has a _load constant_ opcode (`nir_intrinsic_load_constant`), which is used to "load" the sentinel value into some virtual register, and that the myriad of optimization passes NIR is run through can optimize it out (if it's possible to pre-compute some ALU operation or however else take a static decision about the fate of that value, for instance.)
+
+The solution was simply to invent a new opcode (`nir_intrinsic_load_constant_non_opt`), which works exactly like the original one, but, by being in the end another one, is unkown to the optimization passes, which have no option but leaving it in place.
+
+### 3. Obtaining patchable DXIL
+
+The last thing we ask Mesa to do for us is generating DXIL from the nicely patched NIR we have at the moment. But our "deliverable" consists also of a table of the bit offsets to patch in the DXIL for each SC id.
+
+I patched the code that emits constants to the DXIL stream so that it recognizes cases of SCs if they match the sentinel value mask. The rest of the bits are the SC id. On each occurrence, the corresponding callback is leveraged to provide the information to the Godot D3D12 driver. (It's a bit more complicated than that, since the offset in the bitcode has later to be adjusted to match where the `DXBC` chunk ends up in the whole DXIL blob.)
+
+### 4. Creating pipelines with patched DXIL
+
+The Direct3D rendering driver is now able to apply whatever specialization constant values are needed to the already compiled shader, which was our initial goal.
+
+Now I have to mention that LLVM bitcode is an akward beast and that it's not trivial to patch a miserable integer due to variable bit-rate encoding. (Speaking of that, the former approach I implemented had a limit on the number of bits that could be patched, which imposed big limitations in the usable range of integers and imposed an elephant-sized epsilon for floating point ones; those have been happily lifted in the current one.)
+
+This is the comparison of the assembly code produced by the same shader, only with different SC values, which allowed in the second case to optimize out a multiplication:
+![Sample ISA code A](/storage/blog/d3d12/isa-a.png)
+![Sample ISA code B](/storage/blog/d3d12/isa-b.png)
+
+Microsoft may very well end up adding SCs natively to Direct3D. Nonetheless, in the current state of things this hack/technique/magic is **very close to what real SCs would be in terms of performance**. The ISA assembly takes the patched values into account so it can run its own informed optimization passes and so provide a late form of what we were making an effort to prevent at earlier stages. Maybe the biggest downside is that the DXIL blob has to be validated-signed every time it's patched. which takes time (not much, though) and forces us to include the `dxil.dll` file with the binary distribution of the engine.
```

Comment:
```suggestion
Microsoft may very well end up adding SCs natively to Direct3D. Nonetheless, in the current state of things this hack/technique/magic is **very close to what real SCs would be in terms of performance**. The ISA assembly takes the patched values into account so it can run its own informed optimization passes. This way, it can provide a late form of what we were making an effort to prevent at earlier stages. Maybe the biggest downside is that the DXIL blob has to be validated-signed every time it's patched. This takes some time and forces us to include the `dxil.dll` file with the binary distribution of the engine.
```